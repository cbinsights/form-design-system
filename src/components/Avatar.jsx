import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from '../util/baseElement';

const VALID_LIGHT_COLORS = ['white', 'haze', 'lightGray'];
const VALID_DARK_COLORS = ['orange', 'charcoal', 'navy', 'aqua', 'gray', 'red', 'purple'];
export const VALID_BG_COLORS = [...VALID_LIGHT_COLORS, ...VALID_DARK_COLORS];
export const SIZES = ['s', 'm'];

export const grabInitials = (str) =>
  str
    .split(' ')
    .slice(0, 2)
    .reduce((prev, curr) => prev + curr.charAt(0), '')
    .toUpperCase();

const Avatar = ({ bgColor, imgUrl, size, name, Link, ...rest }) => {
  const Element = baseElement({ href: rest.href, onClick: rest.onClick, as: Link });
  return (
    <Element
      {...rest}
      role="img"
      title={name}
      className={cx('fdsAvatar', `bgColor--${bgColor}`, `fdsAvatar--${size}`, {
        'border--focus': Element !== 'div',
        'color--white': VALID_DARK_COLORS.includes(bgColor),
      })}
    >
      {imgUrl && (
        <span className="fdsAvatar-img" style={{ backgroundImage: `url(${imgUrl})` }} />
      )}
      {name && grabInitials(name)}
    </Element>
  );
};

Avatar.defaultProps = {
  bgColor: 'purple',
  size: 'm',
};

Avatar.propTypes = {
  /** Controls color of button */
  bgColor: PropTypes.oneOf(VALID_BG_COLORS),
  /** Controls the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Control initials displayed, and also used for title accessibility attribute (uses 2 words max for initials) */
  name: PropTypes.string.isRequired,
  /** Sets background image over initials */
  imgUrl: PropTypes.string,
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
};

export default Avatar;
