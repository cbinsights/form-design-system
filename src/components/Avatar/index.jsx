import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from 'util/baseElement';

const LIGHT_COLORS = ['white', 'haze', 'lightGray'];
const DARK_COLORS = ['orange', 'charcoal', 'navy', 'aqua', 'gray', 'red', 'purple'];
export const RADII = ['square', 'circle'];
export const BG_COLORS = [...LIGHT_COLORS, ...DARK_COLORS];
export const SIZES = ['s', 'm'];
export const INITIALS_LENGTH = [1, 2];

export const grabInitials = (str, initialsCount = 2) =>
  str
    .split(' ')
    .slice(0, initialsCount)
    .reduce((prev, curr) => prev + curr.charAt(0), '')
    .toUpperCase();

const Avatar = ({
  bgColor = 'purple',
  size = 'm',
  radius = 'circle',
  'aria-label': ariaLabel = 'Avatar',
  initialsLength = 2,
  imgUrl,
  name,
  Link,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, onClick: rest.onClick, as: Link });
  return (
    <Element
      {...rest}
      role={Element === 'div' ? 'img' : undefined}
      aria-label={ariaLabel}
      title={name}
      className={cx(
        'fdsAvatar',
        `bgColor--${bgColor}`,
        `fdsAvatar--${size}`,
        `fdsAvatar--${radius}`,
        {
          'border--focus': Element !== 'div',
          'color--white': DARK_COLORS.includes(bgColor),
        }
      )}
    >
      {imgUrl && (
        <span className="fdsAvatar-img" style={{ backgroundImage: `url(${imgUrl})` }} />
      )}
      {name && grabInitials(name, initialsLength)}
    </Element>
  );
};

Avatar.propTypes = {
  /** Controls color of button */
  bgColor: PropTypes.oneOf(BG_COLORS),
  /** Controls the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Controls radius of button (slightly rounded square, or circle) */
  radius: PropTypes.oneOf(RADII),
  /** Control initials displayed, and also used for title accessibility attribute (uses 2 words max for initials) */
  name: PropTypes.string.isRequired,
  /** Sets background image over initials */
  imgUrl: PropTypes.string,
  /** Controls how many intials can be displayed */
  initialsLength: PropTypes.oneOf(INITIALS_LENGTH),
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
  /** Adds an aria-label to the component (we add a default one for you, but it's preferred to add one if you can.) */
  'aria-label': PropTypes.string,
};

export default Avatar;
