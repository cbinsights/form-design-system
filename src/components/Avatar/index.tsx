import React, { forwardRef, HTMLAttributes } from 'react';
import * as RadixAvatar from '@radix-ui/react-avatar';
import cc from 'classcat';
import baseElement from 'util/baseElement';
import Icon from 'components/Icon';

const LIGHT_COLORS = ['white', 'haze', 'lightGray'] as const;
const DARK_COLORS = [
  'orange',
  'charcoal',
  'navy',
  'aqua',
  'gray',
  'red',
  'purple',
  'blue',
] as const;
export const RADII = ['square', 'circle'] as const;
export const BG_COLORS = [...LIGHT_COLORS, ...DARK_COLORS] as const;
export const SIZES = ['s', 'm', 'l'] as const;
export const INITIALS_LENGTH = [1, 2] as const;

export const grabInitials = (str: string, initialsCount = 2): string =>
  str
    .split(' ')
    .slice(0, initialsCount)
    .reduce((prev, curr) => prev + curr.charAt(0), '')
    .toUpperCase();

export const trimName = (str: string): string | null => {
  // We want to reject strings which only contain whitespaces
  const regex = /\S/;
  // If it doesn't only contain whitespaces, return a trimmed string,
  // otherwise return null
  return str && regex.test(str) ? str.trim() : null;
};

export interface AvatarProps
  extends HTMLAttributes<HTMLElement | HTMLButtonElement | HTMLAnchorElement> {
  /** Controls color of button */
  bgColor?: typeof BG_COLORS[number];
  /** Controls the size of the button */
  size?: typeof SIZES[number];
  /** Controls radius of button (slightly rounded square, or circle) */
  radius?: typeof RADII[number];
  /** Control initials displayed, and also used for title accessibility attribute (uses 2 words max for initials) */
  name?: string;
  /** Sets background image over initials */
  imgUrl?: string;
  /** Controls how many initials can be displayed */
  initialsLength?: typeof INITIALS_LENGTH[number];
  /**  Accepts FDS Icon. Renders if name or imgUrl are not present  */
  PlaceholderIcon?: Icon;
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link?: React.ComponentType;
  /** Adds an aria-label to the component (we add a default one for you, but it's preferred to add one if you can.) */
  'aria-label'?: string;

  href?: string;
}

const Avatar = forwardRef(
  (
    {
      bgColor = 'purple',
      size = 'm',
      radius = 'circle',
      'aria-label': ariaLabel = 'Avatar',
      initialsLength = 2,
      imgUrl,
      name = '',
      Link,
      PlaceholderIcon,
      ...rest
    }: AvatarProps,
    _ref
  ) => {
    const cleanName = trimName(name);

    const Element = baseElement({ href: rest.href, onClick: rest.onClick, as: Link });

    const placeholderIconSize = () => {
      switch (size) {
        case 's':
          return 16;
        case 'm':
          return 18;
        case 'l':
          return 24;
        default:
          return 16;
      }
    };
    console.log(Element);
    return (
      <RadixAvatar.Root
        {...rest}
        // ref={ref}
        role={Element === 'div' ? 'img' : undefined}
        aria-label={ariaLabel}
        title={cleanName || 'Placeholder Avatar'}
        className={cc([
          {
            'border--focus': Element !== 'div',
            'color--white': (DARK_COLORS as ReadonlyArray<string>).includes(bgColor),
          },
          'fdsAvatar',
          'alignChild--center--center',
          `bgColor--${bgColor}`,
          `fdsAvatar--${size}`,
          `fdsAvatar--${radius}`,
        ])}
      >
        <RadixAvatar.Image className="fdsAvatar-img" src={imgUrl} />
        {cleanName && grabInitials(cleanName, initialsLength)}
        <RadixAvatar.Fallback>
          {!(cleanName || imgUrl) && PlaceholderIcon && (
            <PlaceholderIcon customSize={placeholderIconSize()} />
          )}
        </RadixAvatar.Fallback>
      </RadixAvatar.Root>
    );
  }
);

Avatar.displayName = 'Avatar';

export default Avatar;
