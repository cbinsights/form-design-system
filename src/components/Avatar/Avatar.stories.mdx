import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, text, optionsKnob as options } from '@storybook/addon-knobs';
import { arrayToOptions, StoryWrapper, StoryItem, ImportPath } from 'components/util/storybook';
import { action } from '@storybook/addon-actions';

import Avatar, { BG_COLORS, SIZES, RADII, INITIALS_LENGTH } from '.';

<Meta title="Components/Avatar" decorators={[withKnobs]} component={Avatar} />

# Avatar

<ImportPath component={Avatar} section="media" />

<Story name="Knobs" parameters={{ docs: { disable: true }}}>
  <StoryWrapper>
    <StoryItem>
      <Avatar
        bgColor={options('bgColor', arrayToOptions(BG_COLORS), undefined, {
          display: 'inline-radio',
        })}
        radius={options('radius', arrayToOptions(RADII), undefined, {
          display: 'inline-radio',
        })}
        size={options('size', arrayToOptions(SIZES), undefined, {
          display: 'inline-radio',
        })}
        name={text('name', 'Samuel Jackson')}
        imgUrl={text('imgUrl', undefined)}
        initialsLength={options('initials', arrayToOptions(INITIALS_LENGTH), undefined, {
          display: 'inline-radio',
        })}
      />
    </StoryItem>
    <StoryItem>
      <Avatar
        bgColor={options('bgColor', arrayToOptions(BG_COLORS), undefined, {
          display: 'inline-radio',
        })}
        radius={options('radius', arrayToOptions(RADII), undefined, {
          display: 'inline-radio',
        })}
        size={options('size', arrayToOptions(SIZES), undefined, {
          display: 'inline-radio',
        })}
        name={text('name', 'Samuel Jackson')}
        imgUrl="http://placeimg.com/80/81/animals"
      />
    </StoryItem>
  </StoryWrapper>
</Story>

<Preview isExpanded>
  <Avatar name='Samuel Jackson' />
</Preview>


_Displaying an Avatar with an image, and without an image, for visualization purposes_

<Props of={Avatar} />

### Product Cases

_Pay note to `initialsLength` in these cases._

<Preview>
  <Story name="Product Cases">
    <StoryWrapper>
      <StoryItem>
        <p>Entity Avatar</p>
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" bgColor="navy" initialsLength={1} radius="square" />
      </StoryItem>
      <StoryItem>
        <p>User Avatar</p>
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>

### Colors
Accepted colors are based off FDS `bgColor--{color}` classes: https://cbinsights.github.io/form-design-system/fds-styles/#background-colors

<Preview>
  <Story name="Colors">
    <StoryWrapper>
      <StoryItem>
        <Avatar name="default" />
      </StoryItem>
      {BG_COLORS.map(bgColor => (
        <StoryItem>
          <Avatar name={`${bgColor.slice(0,1)} ${bgColor.slice(1,2)}`} bgColor={bgColor} />
        </StoryItem>
      ))}
      <br />
    </StoryWrapper>
  </Story>
</Preview>

### Images
Initials will show either if you don't pass an image, the image doesn't load, or it hasn't loaded yet.

<Preview>
  <Story name="Variations">
    <StoryWrapper>
      <StoryItem>
        <Avatar name="Hello Impecable" imgUrl="http://placeimg.com/80/81/animals" />
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" imgUrl="invalid url" />
      </StoryItem>
      <br />
    </StoryWrapper>
  </Story>
</Preview>

### Sizes

<Preview>
  <Story name="Sizes">
    <StoryWrapper>
      {SIZES.map(size => (
        <StoryItem>
          <Avatar name="Hello Impecable " size={size} />
        </StoryItem>
      ))}
      <br />
    </StoryWrapper>
  </Story>
</Preview>

### Radii

<Preview>
  <Story name="Radii">
    <StoryWrapper>
      {RADII.map(radius => (
        <StoryItem>
          <Avatar name="Hello Impecable " radius={radius} />
        </StoryItem>
      ))}
    </StoryWrapper>
  </Story>
</Preview>

### Button

Avatar will render as a button if you pass it a href, onClick, or Link to it.

<Preview>
  <Story name="Button">
    <StoryWrapper>
      <StoryItem>
        <Avatar name="Hello Impecable" href="#" />
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" onClick={action('avatar-click')} />
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" imgUrl="http://placeimg.com/80/81/animals" href="#" />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>

