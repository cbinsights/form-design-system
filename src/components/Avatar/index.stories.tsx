import React from 'react';
import { StoryObj } from '@storybook/react';
import StarFilledIcon from 'icons/react/StarFilledIcon';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';
import Avatar, { AvatarProps, BG_COLORS, SIZES, RADII } from '.';

export const Primary: StoryObj<AvatarProps> = {
  render: (args: AvatarProps) => (
    <StoryWrapper>
      <StoryItem>
        <Avatar {...args} />
      </StoryItem>
      <StoryItem>
        <Avatar {...args} imgUrl="http://placeimg.com/80/81/animals" />
      </StoryItem>
    </StoryWrapper>
  ),
  args: {
    name: 'Samuel Jackson',
  },
};

export const PlaceholderIcon: StoryObj<AvatarProps> = {
  render: (args: AvatarProps) => (
    <StoryWrapper>
      <StoryItem>
        <Avatar {...args} />
      </StoryItem>
      <StoryItem>
        <Avatar {...args} imgUrl="http://placeimg.com/80/81/animals" />
      </StoryItem>
    </StoryWrapper>
  ),
  args: {
    PlaceholderIcon: StarFilledIcon,
  },
  parameters: {
    docs: {
      description: {
        story: 'Renders if `name` and `imgUrl` are not present.',
      },
    },
  },
};

export const Colors: StoryObj<AvatarProps> = {
  render: (args: AvatarProps): JSX.Element => (
    <StoryWrapper>
      <StoryItem>
        <Avatar {...args} name="default" />
      </StoryItem>
      {BG_COLORS.map((bgColor, idx) => (
        <StoryItem key={idx}>
          <Avatar
            {...args}
            name={`${bgColor.slice(0, 1)} ${bgColor.slice(1, 2)}`}
            bgColor={bgColor}
          />
        </StoryItem>
      ))}
    </StoryWrapper>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Accepted colors are based off [background color classes](?path=/docs/style-utilities-background--background-colors).',
      },
    },
  },
};

export const Images: StoryObj<AvatarProps> = {
  render: (args: AvatarProps): JSX.Element => (
    <StoryWrapper>
      <StoryItem>
        <Avatar
          {...args}
          name="Hello Impecable"
          imgUrl="http://placeimg.com/80/81/animals"
        />
      </StoryItem>
      <StoryItem>
        <Avatar {...args} name="Hello Impecable" imgUrl="invalid url" />
      </StoryItem>
    </StoryWrapper>
  ),
  parameters: {
    docs: {
      description: {
        story:
          "Initials will show either if you don't pass an image, the image doesn't load, or it hasn't loaded yet.",
      },
    },
  },
};

export const Sizes: StoryObj<AvatarProps> = {
  render: (args: AvatarProps): JSX.Element => (
    <StoryWrapper>
      {SIZES.map((size, idx) => (
        <StoryItem key={idx}>
          <Avatar {...args} name="Hello Impecable " size={size} />
        </StoryItem>
      ))}
    </StoryWrapper>
  ),
};

export const Radii: StoryObj<AvatarProps> = {
  render: (args: AvatarProps): JSX.Element => (
    <StoryWrapper>
      {RADII.map((radius, idx) => (
        <StoryItem key={idx}>
          <Avatar {...args} name="Hello Impecable " radius={radius} />
        </StoryItem>
      ))}
    </StoryWrapper>
  ),
};

export const Button: StoryObj<AvatarProps> = {
  render: (args: AvatarProps): JSX.Element => (
    <StoryWrapper>
      <StoryItem>
        <Avatar {...args} name="Hello Impecable" href="#" />
      </StoryItem>
      <StoryItem>
        <Avatar {...args} name="Hello Impecable" />
      </StoryItem>
      <StoryItem>
        <Avatar
          {...args}
          name="Hello Impecable"
          imgUrl="http://placeimg.com/80/81/animals"
          href="#"
        />
      </StoryItem>
    </StoryWrapper>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Avatar will render as a button if you pass it a href, onClick, or Link to it.',
      },
    },
  },
};

export default {
  component: Avatar,
  title: 'components/Avatar',
  parameters: {
    docs: {
      fds: {
        attributes: ['rest'],
        related: ['IconButton', 'StackedButton'],
      },
    },
  },
};
