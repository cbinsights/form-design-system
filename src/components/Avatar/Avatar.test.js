import React from 'react';
import { shallow } from 'enzyme';

import Avatar, { grabInitials } from '.';

const renderComponent = (props) => shallow(<Avatar {...props} />);

describe('grabInitials', () => {

  it('Outputs one initial for one word', () => {
    expect(grabInitials('George')).toBe('G');
  });

  it('Outputs two initials for two words', () => {
    expect(grabInitials('George Wallace')).toBe('GW');
  });

  it('Always outputs uppercase initials', () => {
    expect(grabInitials('george wallace')).toBe('GW');
  });

  it('Only ever outputs the first two initials', () => {
    expect(grabInitials('George Wallace Kennedy')).toBe('GW');
  });

  it('Only outputs one initial when we pass initialsCount of 1 in', () => {
    expect(grabInitials('George Wallace Kennedy', 1)).toBe('G');
  });

});

describe('Avatar component', () => {

  it('has no image', () => {
    const component = renderComponent({ name: 'Joey Tribbiani' });
    expect(component).toMatchSnapshot();
  });

  it('has an image', () => {
    const component = renderComponent({ name: 'Joey Tribbiani', imgUrl: 'lol.png' });
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = renderComponent({ name: 'Ross Gellar', bgColor: 'aqua', size: 's', imgUrl: 'lol.png' })
    expect(component).toMatchSnapshot();
  });
});
