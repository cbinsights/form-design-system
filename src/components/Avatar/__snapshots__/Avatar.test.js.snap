// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component has an image 1`] = `
<div
  aria-label="Avatar"
  className="fdsAvatar bgColor--purple fdsAvatar--m fdsAvatar--circle color--white"
  role="img"
  title="Joey Tribbiani"
>
  <span
    className="fdsAvatar-img"
    style={
      Object {
        "backgroundImage": "url(lol.png)",
      }
    }
  />
  JT
</div>
`;

exports[`Avatar component has no image 1`] = `
<div
  aria-label="Avatar"
  className="fdsAvatar bgColor--purple fdsAvatar--m fdsAvatar--circle color--white"
  role="img"
  title="Joey Tribbiani"
>
  JT
</div>
`;

exports[`Avatar component matches snapshot (set all props) 1`] = `
<div
  aria-label="Avatar"
  className="fdsAvatar bgColor--aqua fdsAvatar--s fdsAvatar--circle color--white"
  role="img"
  title="Ross Gellar"
>
  <span
    className="fdsAvatar-img"
    style={
      Object {
        "backgroundImage": "url(lol.png)",
      }
    }
  />
  RG
</div>
`;
